### PR TITLE
test: make test_sending_fail_with_net_error more statable

### LINKED ISSUE #11365: `test_sending_fail_with_net_error` unstable
## Development Task
test `test_sending_fail_with_net_error` unstable, see details:
https://ci.pingcap.net/blue/organizations/jenkins/tikv_ghpr_test/detail/tikv_ghpr_test/11112/pipeline

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?
Issue Number: close  #11365

Problem Summary:

### What is changed and how it works?

replace channel with sleep

What's Changed:

### Check List 

Tests 

- Unit test

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: tests/failpoints/cases/test_snap.rs [hunk 1]
```diff
@@ -545,24 +545,33 @@ fn test_sending_fail_with_net_error() {
...
+        )),
+    );
 
-    // store2 will not receive any data from stream
-    fail::cfg(on_send_store_fp, "return()").unwrap();
+    // store2 will return err when receive data
+    fail::cfg("receiving_snapshot_net_error", "return()").unwrap();
     pd_client.add_peer(1, new_learner_peer(2, 2));
 
+    // We are ready to recv notify.
```
Comment (nolouch) [inline tests/failpoints/cases/test_snap.rs:569]:
```suggestion
    // ready to recv notify.
```

File: tests/failpoints/cases/test_snap.rs [hunk 2]
```diff
@@ -545,24 +545,33 @@ fn test_sending_fail_with_net_error() {
...
+        Box::new(MessageTypeNotifier::new(
+            MessageType::MsgSnapshot,
+            notify_tx,
+            Arc::clone(&ready_notify),
+        )),
+    );
 
-    // store2 will not receive any data from stream
-    fail::cfg(on_send_store_fp, "return()").unwrap();
+    // store2 will return err when receive data
```
Comment (nolouch) [inline tests/failpoints/cases/test_snap.rs:565]:
```suggestion
    // store2 will return err when receive data.
```

**Review by nolouch — COMMENTED** [review]
rest lgtm

**Review by nolouch — APPROVED** [review]

File: tests/failpoints/cases/test_snap.rs [hunk 3]
```diff
@@ -545,24 +545,33 @@ fn test_sending_fail_with_net_error() {
     let mut cluster = new_server_cluster(1, 2);
     configure_for_snapshot(&mut cluster);
 
-    let on_send_store_fp = "receiving_snapshot_net_error";
```
Comment (lhy1024) [inline tests/failpoints/cases/test_snap.rs:548]: why remove it？

Comment (bufferflies) [reply tests/failpoints/cases/test_snap.rs:548]: not move ,only use string  literal at line 566

**Review by lhy1024 — APPROVED** [review]

**Comment by BusyJay** [comment]
The race condition still exists. assert and receive failure can happen in arbitrary order.

**Comment by BusyJay** [comment]
I tested this PR locally, it failed the 20th time.

**Comment by bufferflies** [comment]
> I tested this PR locally, it failed the 20th time.

it will add receive filter to get the finish singal

File: tests/failpoints/cases/test_snap.rs [hunk 4]
```diff
@@ -545,24 +545,33 @@ fn test_sending_fail_with_net_error() {
...
     // Disable default max peer count check.
     pd_client.disable_default_operator();
     cluster.run_conf_change();
 
     cluster.must_put(b"k1", b"v1");
+    let ready_notify = Arc::default();
+    let (notify_tx, notify_rx) = mpsc::channel();
+    cluster.sim.write().unwrap().add_send_filter(
+        1,
+        Box::new(MessageTypeNotifier::new(
```
Comment (nolouch) [inline tests/failpoints/cases/test_snap.rs:558]: I think you can use custom filter, that pass the first `MsgSnapshot` message, then drop all `MsgSnapshot` latter.

Comment (bufferflies) [reply tests/failpoints/cases/test_snap.rs]: yes

File: components/test_raftstore/src/transport_simulate.rs [hunk 5]
```diff
@@ -576,6 +577,68 @@ impl Filter for CollectSnapshotFilter {
         }
     }
 }
+/// `SendMsgOnceFilter` is a simulation transport filter to simulate the net error.
+/// it will send/receive message util these message contained the filter_type message
+pub struct SendMsgOnceFilter {
```
Comment (BusyJay) [inline components/test_raftstore/src/transport_simulate.rs:582]: How about `RegionPacketFilter::new(region_id, peer_id).msg_type(MessageType::MsgSnapshot).allow(1).build()`?

File: components/test_raftstore/src/transport_simulate.rs [hunk 6]
```diff
@@ -576,6 +577,68 @@ impl Filter for CollectSnapshotFilter {
...
+                Ok(_) => {
+                    n -= 1;
+                }
+                Err(v) => n = v,
+            }
+        }
+
+        Ok(())
+    }
+}
```
Comment (gengliqi) [inline components/test_raftstore/src/transport_simulate.rs:641]: I don't understand what is the use of these codes. 
Seems `pending_notify` is useless.
Also `if self.once.load(Ordering::SeqCst) {` should be `if !self.once.load(Ordering::SeqCst)`.

File: tests/failpoints/cases/test_snap.rs [hunk 7]
```diff
@@ -545,24 +545,37 @@ fn test_sending_fail_with_net_error() {
...
+    cluster.sim.write().unwrap().add_send_filter(
+        1,
+        Box::new(SendMsgOnceFilter::new(notify_tx, MessageType::MsgSnapshot)),
+    );
+    // should ensure learner can receive the snapshot request
+    let (receive_tx, receive_rx) = mpsc::channel();
+    cluster.sim.write().unwrap().add_recv_filter(
+        2,
+        Box::new(SendMsgOnceFilter::new(receive_tx, MessageType::MsgSnapshot)),
+    );
```
Comment (gengliqi) [inline tests/failpoints/cases/test_snap.rs:566]: If `receiving_snapshot_net_error` failpoint turns on, peer 2 should not receive `MsgSnapshot`.
Normally the code here should be blocked forever.`receive_rx.recv_timeout(Duration::from_secs(3)).unwrap()`
However, as the comment above said, there is a bug in `SendMsgOnceFilter` so the block does not happen.

File: components/test_raftstore/src/transport_simulate.rs (hunk 6 above)
Comment (bufferflies) [reply components/test_raftstore/src/transport_simulate.rs:641]: yes, the condition is wrong

File: tests/failpoints/cases/test_snap.rs (hunk 7 above)
Comment (bufferflies) [reply tests/failpoints/cases/test_snap.rs:566]: the test arim that peer2 will interrupt in receiving snapshot to test the receiving count is right.

**Review by gengliqi — APPROVED** [review]

**Review by BusyJay — APPROVED** [review]

**Comment by nolouch** [comment]
ptal @BusyJay

**Pull request merged**